File: android/src/main/java/expo/modules/appshortcuts/AppShortcutsModule.kt
package expo.modules.appshortcuts

import android.content.Intent
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.drawable.Icon as AndroidIcon
import android.os.Build
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class AppShortcutsModule : Module() {
  private var hasListeners = false
  private var pendingShortcut: ShortcutItemRecord? = null

  override fun definition() = ModuleDefinition {
    Name("AppShortcuts")

    Events("onShortcut")

    OnCreate {
      cacheInitialShortcutFromActivityIntent()
    }

    OnActivityEntersForeground {
      cacheInitialShortcutFromActivityIntent()
    }

    OnStartObserving("onShortcut") {
      hasListeners = true
    }

    OnStopObserving("onShortcut") {
      hasListeners = false
    }

    AsyncFunction("setShortcuts") { items: List<ShortcutItemRecord> ->
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
        return@AsyncFunction
      }

      val context = appContext.reactContext
      if (context == null) {
        return@AsyncFunction
      }

      val shortcutManager = context.getSystemService(ShortcutManager::class.java)
      if (shortcutManager == null) {
        return@AsyncFunction
      }

      val launchIntent = context.packageManager.getLaunchIntentForPackage(context.packageName)
      if (launchIntent == null) {
        return@AsyncFunction
      }

      val maxShortcutCount = shortcutManager.maxShortcutCountPerActivity

      shortcutManager.dynamicShortcuts = items
        .take(maxShortcutCount)
        .map { item -> item.toShortcutInfo(context, launchIntent) }
    }

    AsyncFunction("clearShortcuts") { _: Unit? ->
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
        pendingShortcut = null
        return@AsyncFunction
      }

      val context = appContext.reactContext
      if (context == null) {
        return@AsyncFunction
      }

      val shortcutManager = context.getSystemService(ShortcutManager::class.java)
      if (shortcutManager == null) {
        return@AsyncFunction
      }

      shortcutManager.removeAllDynamicShortcuts()
      pendingShortcut = null
    }

    AsyncFunction("getInitialShortcut") {
      pendingShortcut
    }

    OnNewIntent { intent ->
      val shortcut = intent.toShortcutItemRecord() ?: return@OnNewIntent
      pendingShortcut = shortcut

      if (hasListeners) {
        sendEvent("onShortcut", shortcut.toPayload())
      }
    }

    View<AppShortcutsViewProps>("AppShortcuts") { props ->
      AppShortcutsCard(props)
    }
  }

  private fun cacheInitialShortcutFromActivityIntent() {
    val shortcut = appContext.currentActivity?.intent?.toShortcutItemRecord() ?: return
    pendingShortcut = shortcut
  }
}

private data class ShortcutItemRecord(
  @Field
  val id: String = "",
  @Field
  val title: String = "",
  @Field
  val subtitle: String? = null,
  @Field
  val icon: String? = null,
  @Field
  val url: String? = null,
  @Field
  val params: Map<String, String>? = null
) : Record {
  fun toShortcutInfo(context: android.content.Context, baseIntent: Intent): ShortcutInfo {
    val intent = Intent(baseIntent).apply {
      action = Intent.ACTION_VIEW
      flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP
      putExtra(EXTRA_SHORTCUT_ID, id)
      putExtra(EXTRA_SHORTCUT_TITLE, title)
      putExtra(EXTRA_SHORTCUT_SUBTITLE, subtitle)
      putExtra(EXTRA_SHORTCUT_ICON, icon)
      putExtra(EXTRA_SHORTCUT_URL, url)

      val encodedParams = AppShortcutsSerialization.encodeParams(params)
      if (encodedParams != null) {
        putExtra(EXTRA_SHORTCUT_PARAMS, encodedParams)
      } else {
        removeExtra(EXTRA_SHORTCUT_PARAMS)
      }
    }

    return ShortcutInfo.Builder(context, id)
      .setShortLabel(title)
      .apply {
        if (!subtitle.isNullOrEmpty()) {
          setLongLabel(subtitle)
        }

        AppShortcutsIconResolver.icon(context, icon)?.let(::setIcon)
      }
      .setIntent(intent)
      .build()
  }

  fun toPayload(): Map<String, Any?> {
    return mapOf(
      "id" to id,
      "title" to title,
      "subtitle" to subtitle,
      "icon" to icon,
      "url" to url,
      "params" to params
    )
  }
}

private fun Intent.toShortcutItemRecord(): ShortcutItemRecord? {
  val id = getStringExtra(EXTRA_SHORTCUT_ID) ?: return null

  return ShortcutItemRecord(
    id = id,
    title = getStringExtra(EXTRA_SHORTCUT_TITLE) ?: id,
    subtitle = getStringExtra(EXTRA_SHORTCUT_SUBTITLE),
    icon = getStringExtra(EXTRA_SHORTCUT_ICON),
    url = getStringExtra(EXTRA_SHORTCUT_URL),
    params = AppShortcutsSerialization.decodeParams(getStringExtra(EXTRA_SHORTCUT_PARAMS))
  )
}

internal object AppShortcutsIconResolver {
  private val symbolResourceNames = mapOf(
    "home" to "as_material_symbol_home",
    "inbox" to "as_material_symbol_inbox",
    "edit_square" to "as_material_symbol_edit_square",
    "edit" to "as_material_symbol_edit_square",
    "compose" to "as_material_symbol_edit_square",
    "house" to "as_material_symbol_home",
    "tray.full" to "as_material_symbol_inbox",
    "square.and.pencil" to "as_material_symbol_edit_square"
  )

  fun icon(context: android.content.Context, iconName: String?): AndroidIcon? {
    val drawableId = resourceId(context, iconName)
      ?: context.applicationInfo.icon.takeIf { it != 0 }
      ?: return null

    return AndroidIcon.createWithResource(context, drawableId)
  }

  fun resourceId(context: android.content.Context, iconName: String?): Int? {
    val resources = context.resources
    val packageName = context.packageName
    val resolvedName = iconName?.trim()?.takeIf { it.isNotEmpty() }

    return resolvedName
      ?.let { name -> resourceNameCandidates(name) }
      ?.firstNotNullOfOrNull { candidate ->
        resources.getIdentifier(candidate, "drawable", packageName).takeIf { it != 0 }
          ?: resources.getIdentifier(candidate, "mipmap", packageName).takeIf { it != 0 }
      }
  }

  internal fun resourceNameCandidates(iconName: String): List<String> {
    val normalizedName = iconName
      .lowercase()
      .replace(Regex("[^a-z0-9_]"), "_")
      .trim('_')

    return listOfNotNull(
      iconName,
      normalizedName.takeIf { it.isNotEmpty() },
      normalizedName.takeIf { it.isNotEmpty() }?.let { "as_material_symbol_$it" },
      symbolResourceNames[iconName],
      symbolResourceNames[iconName.lowercase()],
      symbolResourceNames[normalizedName],
      normalizedName.takeIf { it.isNotEmpty() }?.let { "as_shortcut_$it" }
    ).distinct()
  }
}

private const val EXTRA_SHORTCUT_ID = "expo.modules.appshortcuts.extra.ID"
private const val EXTRA_SHORTCUT_TITLE = "expo.modules.appshortcuts.extra.TITLE"
private const val EXTRA_SHORTCUT_SUBTITLE = "expo.modules.appshortcuts.extra.SUBTITLE"
private const val EXTRA_SHORTCUT_ICON = "expo.modules.appshortcuts.extra.ICON"
private const val EXTRA_SHORTCUT_URL = "expo.modules.appshortcuts.extra.URL"
private const val EXTRA_SHORTCUT_PARAMS = "expo.modules.appshortcuts.extra.PARAMS"

File: ios/AppShortcutsModule.swift
import ExpoModulesCore
#if os(iOS)
import UIKit
#endif

public class AppShortcutsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AppShortcuts")

    Events("onShortcut")

    OnCreate {
      #if os(iOS)
      AppShortcutsRegistry.shared.attach(module: self)
      #endif
    }

    OnDestroy {
      #if os(iOS)
      AppShortcutsRegistry.shared.detach(module: self)
      #endif
    }

    OnStartObserving("onShortcut") {
      #if os(iOS)
      AppShortcutsRegistry.shared.setHasListeners(true)
      #endif
    }

    OnStopObserving("onShortcut") {
      #if os(iOS)
      AppShortcutsRegistry.shared.setHasListeners(false)
      #endif
    }

    AsyncFunction("setShortcuts") { (items: [ShortcutItemRecord]) in
      #if os(iOS)
      UIApplication.shared.shortcutItems = items.map { $0.toShortcutItem() }
      #endif
    }
    .runOnQueue(.main)

    AsyncFunction("clearShortcuts") {
      #if os(iOS)
      UIApplication.shared.shortcutItems = []
      AppShortcutsRegistry.shared.clearPendingShortcut()
      #endif
    }
    .runOnQueue(.main)

    AsyncFunction("getInitialShortcut") { () -> ShortcutItemRecord? in
      #if os(iOS)
      return AppShortcutsRegistry.shared.pendingShortcut
      #else
      return nil
      #endif
    }

    View(AppShortcutsView.self)
  }
}

struct ShortcutItemRecord: Record {
  @Field
  var id: String = ""

  @Field
  var title: String = ""

  @Field
  var subtitle: String?

  @Field
  var icon: String?

  @Field
  var url: String?

  @Field
  var params: [String: String]?

  func toPayload() -> [String: Any] {
    descriptor.payload
  }

  var descriptor: AppShortcutDescriptor {
    AppShortcutDescriptor(
      id: id,
      title: title,
      subtitle: subtitle,
      icon: icon,
      url: url,
      params: params
    )
  }
}

#if os(iOS)
extension ShortcutItemRecord {
  func toShortcutItem() -> UIApplicationShortcutItem {
    UIApplicationShortcutItem(
      type: id,
      localizedTitle: title,
      localizedSubtitle: subtitle,
      icon: AppShortcutsIconResolver.icon(from: icon),
      userInfo: descriptor.userInfo
    )
  }

  static func from(shortcutItem: UIApplicationShortcutItem) -> ShortcutItemRecord {
    var record = ShortcutItemRecord()
    let descriptor = AppShortcutDescriptor.from(
      userInfo: shortcutItem.userInfo,
      fallbackId: shortcutItem.type,
      fallbackTitle: shortcutItem.localizedTitle,
      fallbackSubtitle: shortcutItem.localizedSubtitle
    )

    record.id = descriptor.id
    record.title = descriptor.title
    record.subtitle = descriptor.subtitle
    record.icon = descriptor.icon
    record.url = descriptor.url
    record.params = descriptor.params

    return record
  }
}

final class AppShortcutsRegistry {
  static let shared = AppShortcutsRegistry()

  private weak var module: AppShortcutsModule?
  private(set) var pendingShortcut: ShortcutItemRecord?
  private var hasListeners = false

  private init() {}

  func attach(module: AppShortcutsModule) {
    self.module = module
  }

  func detach(module: AppShortcutsModule) {
    if self.module === module {
      self.module = nil
    }
  }

  func setHasListeners(_ hasListeners: Bool) {
    self.hasListeners = hasListeners
  }

  func clearPendingShortcut() {
    pendingShortcut = nil
  }

  func storeInitialShortcut(_ shortcutItem: UIApplicationShortcutItem) {
    pendingShortcut = ShortcutItemRecord.from(shortcutItem: shortcutItem)
  }

  func handleShortcut(_ shortcutItem: UIApplicationShortcutItem) {
    let record = ShortcutItemRecord.from(shortcutItem: shortcutItem)
    pendingShortcut = record

    if hasListeners {
      module?.sendEvent("onShortcut", record.toPayload())
    }
  }
}

enum AppShortcutsIconResolver {
  static func icon(from value: String?) -> UIApplicationShortcutIcon? {
    guard let value, !value.isEmpty else {
      return nil
    }

    return UIApplicationShortcutIcon(systemImageName: value)
  }
}
#endif
